keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,617 @@
+will only be updated if a Node addition or removal results in a transition above
+or below the overload threshold. For example, syncs would be triggered in either
+of the following scenarios:
+
+1. A deleted Node results in a Service exceeding the overload threshold.
+2. A new Node results in a Service that is able to achieve an endpoint
+   distribution below 20% for the first time.
+
+### Future Expansion
+
+In the future we may expand this functionality if needed. This could include:
+
+- A new `RequireZone` algorithm that would keep endpoints in EndpointSlices for
+  the same zone they are in.
+- A new option to specify a minimum threshold for the `Auto` approach.
+- Support for region based hints.
+
+### Test Plan
+
+#### Controller Unit Tests
+| Test Description | Expected Result |
+| :--- | :--- |
+| Feature Gate On, Annotation == 'Auto', 2+ zones | Hints set |
+| Feature Gate On, Annotation == 'Auto', 1 zone | No hints set |
+| Feature Gate On, Annotation == 'On', 2+ zones | No hints |
+| Feature Gate On, Annotation Off, 2+ zones | No hints |
+| Feature Gate Off, Annotation On, 2+ zones | No hints |
+| Feature Gate Off, Annotation Off, 2+ zones | No hints |
+| Feature Gate Off, Annotation Off, 2+ zones | No hints |
+| 2 endpoints, 3 zones | No hints |
+| 3 endpoints, 3 zones | Hints set |
+| 4 endpoints, 3 zones | No hints |
+| 4 endpoints, 2 zones | Hints set |
+| 4 endpoints all from 1 zone, 2 zones | Hints set |
+| 4 endpoints, 3 zones, 1 zone with 2x cores | Hints set |
+| 400 endpoints, 4 zones with slightly different cores | Hints set |
+| Node removal that does not trigger threshold transition | No EndpointSlice changes |
+| Node removal that triggers threshold transition | EndpointSlice updates |
+| Node without way to determine cores | All Nodes treated equally |
+| Endpoint additions that require redistribution | Hints updated |
+| Endpoint removals that require redistribution | Hints updated |
+
+#### Kube-Proxy Unit Tests
+| Test Description | Expected Result |
+| :--- | :--- |
+| Feature Gate On, Annotation == 'Auto', hints matching zone | Endpoints filtered |
+| Feature Gate On, Annotation == 'On', hints matching zone | Endpoints not filtered |
+| Feature Gate Off, Annotation == 'Auto', hints matching zone | Endpoints not filtered |
+| Feature Gate On, Annotation == 'Auto', no hints matching zone | Endpoints not filtered |
+
+### Observability
+We can reuse some of the metrics of EndpointSlice Controller that we already
+have in the current version to observe the changes of endpoints (addition,
+deletion and update). Meanwhile we can add more metrics to have a glimpse of
+different approaches.
+
+- `endpoint_slice_controller/endpoint_overload_by_zone`
+- `endpoint_slice_controller/endpointslices_changed_per_sync`
+- `endpoint_slice_controller/syncs`
+
+```
+const SubSystem = "endpoint_slice_controller"
+
+// This metric observes overload per zone for endpoints in EndpointSlices
+EPSOverloadByZone = metrics.NewHistogramVec(
```

Comment:
How are we going to report this metric?
Whenever you process a particular service you report it for each of zones?

I'm definitely not against this metric, but I'm having some troubles with intuitively explaining what it means to myself.

Reply:
Yeah I'm really not sure, I was hoping that reporting every value when a Service syncs would give us somewhat helpful data as far as distribution. I think some visibility into overload would be really helpful but I still haven't found a great way to communicate that with a metric.

Reply:
I see two possible use cases:
- help the users to understand how good or bad distributed are their deployments, i.e they are overloading zoneA, ...
- understand the distribution algorithm convergence, I think that in a dynamic environment this is going to fluctuate, if you see that this is changing constantly that means that you are constantly redistributing endpoints, that will impact workloads

Reply:
Sure - I fully buy the usecases (those are the same ones I had on my mind).

My concern was more: "would this metric allow us answering those questions"?
And I'm not convinced (at least not intuitively...)

Reply:
Yeah, I'm not sure it would either. I've been trying to think of alternatives but haven't so far. To avoid overcommitting to anything, I've removed this from the KEP for now, will keep on thinking of potential metrics we could add here. Any ideas would be very appreciated.